src/meta-srv/src/service/store/txn/etcd.rs

```diff
@@ -0,0 +1,134 @@
+                .collect::<Vec<_>>();
+            etcd_txn = etcd_txn.when(compares);
+        }
+        if txn.c_then {
+            let success = txn
+                .req
+                .success
+                .into_iter()
+                .map(EtcdTxnOp::from)
+                .collect::<Vec<_>>();
+            etcd_txn = etcd_txn.and_then(success);
+        }
+        if txn.c_else {
+            let failure = txn
+                .req
+                .failure
+                .into_iter()
+                .map(EtcdTxnOp::from)
+                .collect::<Vec<_>>();
+            etcd_txn = etcd_txn.or_else(failure);
+        }
+        etcd_txn
+    }
+}
+
+impl From<Compare> for EtcdCompare {
+    fn from(cmp: Compare) -> Self {
+        let etcd_cmp = match cmp.cmp {
+            CompareOp::Equal => EtcdCompareOp::Equal,
+            CompareOp::Greater => EtcdCompareOp::Greater,
+            CompareOp::Less => EtcdCompareOp::Less,
+            CompareOp::NotEqual => EtcdCompareOp::NotEqual,
+        };
+        match cmp.target {
+            Some(target) => EtcdCompare::value(cmp.key, etcd_cmp, target),
+            // create revision 0 means key was not exist
+            None => EtcdCompare::create_revision(cmp.key, etcd_cmp, 0),
+        }
+    }
+}
+
+impl From<TxnOp> for EtcdTxnOp {
+    fn from(op: TxnOp) -> Self {
+        match op {
+            TxnOp::Put(key, value) => EtcdTxnOp::put(key, value, None),
+            TxnOp::Get(key) => EtcdTxnOp::get(key, None),
+            TxnOp::Delete(key) => EtcdTxnOp::delete(key, None),
+        }
+    }
+}
+
+impl From<EtcdTxnOpResponse> for TxnOpResponse {
+    fn from(op_resp: EtcdTxnOpResponse) -> Self {
+        match op_resp {
+            EtcdTxnOpResponse::Put(res) => {
+                let prev_kv = res.prev_key().map(KvPair::from_etcd_kv);
+                let put_res = PutResponse {
+                    prev_kv,
+                    ..Default::default()
+                };
+                TxnOpResponse::ResponsePut(put_res)
+            }
+            EtcdTxnOpResponse::Get(res) => {
+                let kvs = res.kvs().iter().map(KvPair::from_etcd_kv).collect();
+                let range_res = RangeResponse {
+                    kvs,
+                    ..Default::default()
+                };
+                TxnOpResponse::ResponseGet(range_res)
+            }
+            EtcdTxnOpResponse::Delete(res) => {
+                let prev_kvs = res
+                    .prev_kvs()
+                    .iter()
+                    .map(KvPair::from_etcd_kv)
+                    .collect::<Vec<_>>();
+                let delete_res = DeleteRangeResponse {
+                    prev_kvs,
+                    deleted: res.deleted(),
+                    ..Default::default()
+                };
+                TxnOpResponse::ResponseDelete(delete_res)
+            }
+            EtcdTxnOpResponse::Txn(_) => unimplemented!("nested txn is not supported"),
```

Comment:
better returns some error since the `EtcdTxnOpResponse` is something we cannot control

Reply:
Because the request type was restricted, the response type was predictable, but I think your suggestion should be adopted, which can avoid us making mistakes in the future.